[14.0][ADD] ebill_postfinance

TDu opened the pull request.

TDu committed:
Add ebill_postfinance

santostelmo commented on ebill_postfinance/__manifest__.py:
```diff
@@ -0,0 +1,38 @@
+        "l10n_ch_qriban",
+        "queue_job",
+        "sale",
+        "sale_stock",
```
Why we need `sale_stock` to be installed ? This will force the stock module to be installed in the system while some companies don't use stock module at all

TDu replied:
It is installed because the module adds  information about the delivery note https://github.com/OCA/l10n-switzerland/pull/675/files#diff-a2170ebdcc7a37e4c7e59582666f475e8805d9059f2798b1cbb04e951f50f549R47 in the xml generated.

TDu replied:
Split into a new `ebill_postfinance_stock` module the base module should not have any stock related dependencies.

yvaucher commented on ebill_postfinance/models/ebill_postfinance_service.py:
```diff
@@ -0,0 +1,153 @@
+import logging
+import logging.config
+
+# from ..ebilling_postfinance.ebilling_postfinance import ebilling_postfinance
```
To remove?

TDu replied:
done

yvaucher commented on ebill_postfinance/models/ebill_postfinance_service.py:
```diff
@@ -0,0 +1,153 @@
+        return res
+
+    def get_process_protocol_list(self, archive_data=False):
+        # Is this the processing result of an invoice ?
```
Should it translate as a TODO ?

TDu replied:
Questioning the use of this call to the api.
The situation at the moment is, all possible call to the api are implemented, so they can be used and tested.
But they are not all used yet by the module for real.
There is the roadmap file that contains some more information.

yvaucher commented on ebill_postfinance/models/ebill_postfinance_service.py:
```diff
@@ -0,0 +1,153 @@
+        service = self._get_service()
+        res = service.get_registration_protocol_list(archive_data)
+        for registration_protocol in res or {}:
+            self.get_registration_protocol(registration_protocol.CreateDate)
```
the value returned from this function is not used

TDu replied:
This call is related to the contract subscription update through the web service. And not yet implemented, actually when I tested it I could never get any data.
I learn today that the subscription update through the web service is only implemented for b2ebanking or b2c.

yvaucher commented on ebill_postfinance/models/ebill_postfinance_service.py:
```diff
@@ -0,0 +1,153 @@
+        help="Timeout for each HTTP (GET, POST) request in seconds.",
+    )
+
+    def _get_service(self):
```
Does it create a new session each time, can't we reuse a session by storing the current service?

TDu replied:
This is a very good point. And yes I believe there is some improvement to be done on this...

ecino replied:
Maybe decorating the method with `@ormcache` would work?

TDu committed:
Add ebill_postfinance_stock

Remove the dependency on stock in the `ebill_postfinance` module.
And add a new module `ebill_postfinance_stock` that will integrate
in the xml invoice informations related to the deliveries.

This is for Odoo implementation that sell only services and have no
stock to manage. And do not use the `stock` module.

TDu commented:
@yvaucher  Thank you for the review, I will attend them.
For now I extracted the stock related code into `ebill_postfinance_stock` see last commit 17b06e2ea2198e16cfbf9e258124206f06dfc838

TDu commented:
Previous fixup fixes the number of references included for each invoice line in the xml payload.
Adds a payload size in the data, can be useful, knowing that the web service will refuse more than 5MB.

i-vyshnevska commented on ebill_postfinance/models/account_move.py:
```diff
@@ -0,0 +1,139 @@
+    def _export_invoice(self):
+        """Export invoice with the help of account_invoice_export module."""
+        postfinance_method = self.env.ref(
+            "ebill_postfinance.postfinance_transmit_method"
```
```suggestion
            "ebill_postfinance.postfinance_transmit_method", raise_if_not_found=False
```

TDu replied:
If this record is not present, nothing will work on the module. So better to have the error raised IMO
Same for the comment below.

i-vyshnevska commented on ebill_postfinance/models/account_move.py:
```diff
@@ -0,0 +1,139 @@
+        if self.move_type not in ("out_invoice", "out_refund"):
+            return
+        postfinance_method = self.env.ref(
+            "ebill_postfinance.postfinance_transmit_method"
```
```suggestion
            "ebill_postfinance.postfinance_transmit_method", raise_if_not_found=False
```

TDu replied:
If this record is not present, nothing will work on the module. So better to have the error raised IMO

santostelmo commented on ebill_postfinance/views/ebill_postfinance_service.xml:
```diff
@@ -0,0 +1,136 @@
+                        <group>
+                            <field name="biller_id" />
+                            <field name="username" />
+                            <field name="password" />
```
```suggestion
                            <field name="password"  password="True"/>
```
Could you please hide the password value?

TDu replied:
done

simahawk commented on ebill_postfinance/models/ebill_postfinance_invoice_message.py:
```diff
@@ -94,13 +94,13 @@ def _compute_payload_size(self):
+        self.transaction_id = "-".join(
+            [
+                fields.Datetime.now().strftime("%y%m%d%H%M%S"),
+                self.invoice_id.name.replace("/", "").replace("_", ""),
```
I would use slugify to make sure there's no weird char.

simahawk commented:
@TDu what's the overall state here?

TDu commented:
I squashed all previous commit and then the last fixup adds the credit note to be sent to the server.

sebalix commented:
@TDu so we are good to merge this?

ecino commented:
Maybe you could increase the coverage of the tests? Also you can force-push to retrigger the runboat. Is this ready to be merged otherwise?

TDu commented:
@sebalix  The required Python library needed to be on Pypi. This is done.
@ecino  More tests would be nice, but I will improve them when migrating to 16.0
@simahawk  IMO this is ready for merge
@yvaucher Could you update your review

ecino commented on ebill_postfinance/models/account_move.py:
```diff
@@ -0,0 +1,144 @@
+        #     error_log = _("An error of type {} occured.").format(
+        #         values.get("error_type")
+        #     )
+        # activity.note += "<div class='mt16'><p>{}</p></div>".format(error_log)
```
This can be removed if it's not needed.

TDu replied:
Removed.

ecino reviewed (commented):
Should you also add `server-env` and `edi` in file `oca_requirements.txt` ?

TDu commented:
> Should you also add `server-env` and `edi` in file `oca_requirements.txt` ?

With the introduction of the use of PyPi to find depending modules, this is not needed anymore, as far as I know, in fact Travis runs without errors.
Thanks for your review.

yvaucher reviewed (approved):
My remarks for improvements can be addressed later, better to have a version we can migrate to other version than none.

yvaucher commented:
/ocabot merge nobump
